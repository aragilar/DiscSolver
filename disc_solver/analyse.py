# -*- coding: utf-8 -*-
"""
Stuff to analyse solutions
"""

from math import pi

import matplotlib.pyplot as plt

from .constants import KM


def generate_plot(angles, soln, B_norm, v_norm, ρ_norm, **kwargs):
    """
    Generate plot, with enough freedom to be able to format fig
    """
    param_names = [
        {
            "name": "B_r",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
        {
            "name": "B_φ",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
        {
            "name": "B_θ",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
        {
            "name": "v_r",
            "y_label": "Velocity Field (km/s)",
            "normalisation": v_norm / KM,  # km/s
        },
        {
            "name": "v_φ",
            "y_label": "Velocity Field (km/s)",
            "normalisation": v_norm / KM,  # km/s
        },
        {
            "name": "v_θ",
            "y_label": "Velocity Field (km/s)",
            "normalisation": v_norm / KM,  # km/s
        },
        {
            "name": "ρ",
            "y_label": "Density ($g cm^{-3}$)",
            "normalisation": ρ_norm,
            "scale": "log",
        },
        {
            "name": "B_φ_prime",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
    ]

    fig, axes = plt.subplots(nrows=2, ncols=4, tight_layout=True, **kwargs)
    axes.shape = len(param_names)
    for i, settings in enumerate(param_names):
        ax = axes[i]
        ax.plot(
            90 - (angles * 180 / pi),
            soln[:, i] * settings["normalisation"]
        )
        ax.set_xlabel("angle from plane (°)")
        ax.set_ylabel(settings["y_label"])
        ax.set_yscale(settings.get("scale", "linear"))
        ax.set_title(settings["name"])
    return fig
