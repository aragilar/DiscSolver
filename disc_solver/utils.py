# -*- coding: utf-8 -*-
"""
Useful functions
"""

from math import pi, cos, sin

import numpy as np
from matplotlib.ticker import FuncFormatter


def is_supersonic(v, B, rho, sound_speed, mhd_wave_type):
    """
    Checks whether velocity is supersonic.
    """
    v_axis = 1 if v.ndim > 1 else 0
    B_axis = 1 if B.ndim > 1 else 0

    v_sq = np.sum(v**2, axis=v_axis)
    B_sq = np.sum(B**2, axis=B_axis)

    cos_sq_psi = np.sum(
        (v.T**2/v_sq).T * (B.T**2/B_sq).T, axis=max(v_axis, B_axis)
    ) ** 2

    v_a_sq = B_sq / (4*pi*rho)
    if mhd_wave_type == "slow":
        mhd_wave_speed = 1/2 * (
            v_a_sq + sound_speed**2 - np.sqrt(
                (v_a_sq + sound_speed**2)**2 -
                4 * v_a_sq * sound_speed**2 * cos_sq_psi
            )
        )
    elif mhd_wave_type == "alfven":
        mhd_wave_speed = v_a_sq * cos_sq_psi
    elif mhd_wave_type == "fast":
        mhd_wave_speed = 1/2 * (
            v_a_sq + sound_speed**2 + np.sqrt(
                (v_a_sq + sound_speed**2)**2 -
                4 * v_a_sq * sound_speed**2 * cos_sq_psi
            )
        )

    return v_sq > mhd_wave_speed


def cot(angle):
    """
    Computes cot
    """
    if angle == pi:
        return float('-inf')
    elif angle == pi/2:
        return 0
    elif angle == 0:
        return float('inf')
    return cos(angle)/sin(angle)


def better_sci_format(physical_axis):
    """
    Use scientific notation for each tick mark for axis `physical_axis`.
    """
    physical_axis.set_major_formatter(
        FuncFormatter(lambda x, pos: "{:.2e}".format(x).replace("-", "âˆ’"))
    )
