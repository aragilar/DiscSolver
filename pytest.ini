[pytest]
filterwarnings =
    once::DeprecationWarning
    once::PendingDeprecationWarning
