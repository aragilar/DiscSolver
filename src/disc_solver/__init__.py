# -*- coding: utf-8 -*-
"""
Solver for PHD Project
"""
from ._version import version as __version__  # noqa
