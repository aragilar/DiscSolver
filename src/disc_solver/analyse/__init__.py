# -*- coding: utf-8 -*-
"""
Analysis component of Disc Solver
"""
