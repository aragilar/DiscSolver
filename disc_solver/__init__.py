# -*- coding: utf-8 -*-
"""
Solver for PHD Project
"""

__version__ = "0.1"

import argparse
from pprint import pprint
import signal
import tempfile
from types import SimpleNamespace

import arrow
import logbook
from logbook.compat import redirected_warnings, redirected_logging

import numpy as np
import h5py

import matplotlib as mpl
mpl.use("Qt4Agg")
mpl.rcParams["backend.qt4"] = "PySide"
import matplotlib.pyplot as plt

from .analyse import generate_plot, info
from .config import define_conditions, get_input
from .solution import solution

log = logbook.Logger(__name__)


def stack_info(frame):
    """Walk up stack printing line and locals"""
    print('f_code', frame.f_code)
    print('f_locals:')
    pprint(frame.f_locals)
    print()
    if frame.f_back:
        stack_info(frame.f_back)


def sig_handler(signum, frame):
    """Custom signal handler"""
    print(signum)
    stack_info(frame)


def solution_main(output_file=None, get_config_file=True):
    """
    Main function to generate solution
    """
    inp, = get_input(get_file=get_config_file)
    if output_file:
        inp.output_file = output_file
    cons = define_conditions(inp)

    signal.signal(signal.SIGINT, sig_handler)

    null_handler = logbook.NullHandler()
    with null_handler.applicationbound():
        angles, soln = solution(
            cons.angles, cons.init_con, inp.β, cons.c_s, cons.norm_kepler_sq,
            cons.η_O, cons.η_A, cons.η_H, max_steps=inp.max_steps,
            taylor_stop_angle=inp.taylor_stop_angle
        )

    with h5py.File(inp.output_file) as f:
        grp = f.create_group(str(arrow.now()))
        grp['angles'] = angles
        grp['solution'] = soln
        grp.attrs.update(vars(inp))


def analyse_main(output_file=None, **kwargs):
    """
    Main function to analyse solution
    """
    if output_file is None:
        output_file, kwargs = analyse_parser(kwargs)
    make_plot = kwargs.get("show") or kwargs.get("output_fig_file")

    with redirected_warnings(), redirected_logging():
        with h5py.File(output_file) as f:
            for grp in f.values():
                inp = SimpleNamespace(**grp.attrs)  # pylint: disable=star-args
                cons = define_conditions(inp)
                angles = np.array(grp["angles"])
                soln = np.array(grp["solution"])
                if make_plot:
                    fig = generate_plot(angles, soln, inp, cons)
                if kwargs.get("show"):
                    plt.show()
                if kwargs.get("output_fig_file"):
                    fig.savefig(kwargs.get("output_fig_file"))
                if kwargs.get("info"):
                    info(inp, cons)


def analyse_parser(kwargs):
    """
    CLI Parser for analyse
    """
    parser = argparse.ArgumentParser(description='Analyser for DiscSolver')
    parser.add_argument("output_file")

    if not kwargs:
        parser.add_argument(
            "--show", "-s", action="store_true", default=False
        )
        parser.add_argument("--output_fig_file", "-o")
        parser.add_argument(
            "--info", "-i", action="store_true", default=False
        )

    output_file = parser.parse_args().output_file

    if not kwargs:
        kwargs = vars(parser.parse_args())
    return output_file, kwargs


def main():
    """
    The main function
    """
    null_handler = logbook.NullHandler()
    with null_handler.applicationbound():
        with tempfile.NamedTemporaryFile() as output_file:
            solution_main(output_file=output_file.name, get_config_file=False)
            analyse_main(
                output_file=output_file.name, show=True,
                output_fig_file="plot.png"
            )
