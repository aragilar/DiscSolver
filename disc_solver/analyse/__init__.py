# -*- coding: utf-8 -*-
"""
Stuff to analyse solutions
"""

from math import pi, sqrt

import numpy as np
import matplotlib.pyplot as plt

from ..constants import KM
from ..utils import better_sci_format, mhd_wave_speeds, MHD_WAVE_INDEX


def generate_plot(angles, soln, inp, cons, **kwargs):
    """
    Generate plot, with enough freedom to be able to format fig
    """
    B_norm, v_norm, ρ_norm = cons.B_norm, cons.v_norm, cons.ρ_norm
    zero_soln = np.zeros(len(soln))
    v = np.array([zero_soln, zero_soln, soln[:, 5]])
    wave_speeds = np.sqrt(mhd_wave_speeds(
        v.T, soln[:, 0:3], soln[:, 6], inp.c_s
    ))
    linestyle = kwargs.pop("line style")
    with_slow = kwargs.pop("with slow")
    with_alfven = kwargs.pop("with alfven")
    with_fast = kwargs.pop("with fast")

    param_names = [
        {
            "name": "B_r",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
        {
            "name": "B_φ",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
        {
            "name": "B_θ",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
        {
            "name": "v_r",
            "y_label": "Velocity Field (km/s)",
            "normalisation": v_norm / KM,  # km/s
        },
        {
            "name": "v_φ",
            "y_label": "Velocity Field (km/s)",
            "normalisation": v_norm / KM,  # km/s
            "offset": sqrt(cons.norm_kepler_sq) * v_norm / KM
        },
        {
            "name": "v_θ",
            "y_label": "Velocity Field (km/s)",
            "normalisation": v_norm / KM,  # km/s
            "legend": True,
            "scale": kwargs.pop("v_θ scale", "linear"),
            "extras": []
        },
        {
            "name": "ρ",
            "y_label": "Density ($g cm^{-3}$)",
            "normalisation": ρ_norm,
            "scale": "log",
        },
        {
            "name": "B_φ_prime",
            "y_label": "Magnetic Field (Gauss)",
            "normalisation": B_norm,
        },
    ]

    if with_slow:
        param_names[5]["extras"].append({
            "label": "slow",
            "data": wave_speeds[MHD_WAVE_INDEX["slow"]],
            "normalisation": v_norm / KM,
        })
    if with_alfven:
        param_names[5]["extras"].append({
            "label": "alfven",
            "data": wave_speeds[MHD_WAVE_INDEX["alfven"]],
            "normalisation": v_norm / KM,
        })
    if with_fast:
        param_names[5]["extras"].append({
            "label": "fast",
            "data": wave_speeds[MHD_WAVE_INDEX["fast"]],
            "normalisation": v_norm / KM,
        })

    fig, axes = plt.subplots(
        nrows=2, ncols=4, tight_layout=True, sharex=True, **kwargs
    )
    axes.shape = len(param_names)
    for i, settings in enumerate(param_names):
        ax = axes[i]
        ax.plot(
            90 - (angles * 180 / pi),
            (
                soln[:, i] * settings["normalisation"] -
                settings.get("offset", 0)
            ), linestyle,
        )
        for extra in settings.get("extras", []):
            ax.plot(
                90 - (angles * 180 / pi),
                extra["data"] * extra["normalisation"],
                label=extra.get("label")
            )
        ax.set_xlabel("angle from plane (°)")
        ax.set_ylabel(settings["y_label"])
        ax.set_yscale(settings.get("scale", "linear"))
        ax.set_title(settings["name"])
        if settings.get("legend"):
            ax.legend()
        better_sci_format(ax.yaxis)
    return fig


def plot_options(parser):
    """
    Add cli arguments for defining plot
    """
    parser.add_argument("--v_θ", choices=("log", "linear"), default="linear")
    parser.add_argument("--line-style", default="-")
    parser.add_argument(
        "--with-slow", action='store_true', default=False)
    parser.add_argument(
        "--with-alfven", action='store_true', default=False)
    parser.add_argument(
        "--with-fast", action='store_true', default=False)


def get_plot_args(args):
    """
    Parse plot args
    """
    return {
        "v_θ scale": args.get("v_θ", "linear"),
        "with slow": args.get("with_slow", False),
        "with alfven": args.get("with_alfven", False),
        "with fast": args.get("with_fast", False),
        "line style": args.get("line_style", "-")
    }


def generate_deriv_plot(angles, soln, inp, cons, **kwargs):
    """
    Generate plot of derivatives
    """
    # pylint: disable=unused-argument
    param_names = [
        {
            "name": "B_r",
        },
        {
            "name": "B_φ",
        },
        {
            "name": "B_θ",
        },
        {
            "name": "v_r",
        },
        {
            "name": "v_φ",
        },
        {
            "name": "v_θ",
        },
        {
            "name": "ρ",
        },
        {
            "name": "B_φ_prime",
        },
    ]

    linestyle = kwargs.pop("line style")
    internal_data = kwargs.pop("internal_data")
    deriv_angles = np.array(internal_data["angles"])
    derivs = np.array(internal_data["derivs"])
    npnot = np.logical_not

    fig, axes = plt.subplots(
        nrows=2, ncols=4, tight_layout=True, sharex=True, **kwargs
    )
    axes.shape = len(param_names)
    for i, settings in enumerate(param_names):
        ax = axes[i]
        pos_deriv = derivs[:, i] >= 0
        ax.plot(
            90 - (deriv_angles[pos_deriv] * 180 / pi),
            derivs[pos_deriv, i], linestyle,
        )
        ax.plot(
            90 - (deriv_angles[npnot(pos_deriv)] * 180 / pi),
            - derivs[npnot(pos_deriv), i], linestyle,
        )
        ax.set_xlabel("angle from plane (°)")
        ax.set_yscale(settings.get("scale", "log"))
        ax.set_title(settings["name"])
        if settings.get("legend"):
            ax.legend()
        better_sci_format(ax.yaxis)
    return fig


def deriv_plot_options(parser):
    """
    Add cli arguments for defining derivative plot
    """
    parser.add_argument("--line-style", default=".")


def get_deriv_plot_args(args):
    """
    Parse plot args
    """
    return {
        "line style": args.get("line_style", "-"),
        "internal_data": args["internal_data"]
    }


def generate_params_plot(angles, soln, inp, cons, **kwargs):
    """
    Generate plot of all values, including intermediate values
    """
    # pylint: disable=unused-argument
    param_names = [
        {
            "name": "B_r",
        },
        {
            "name": "B_φ",
        },
        {
            "name": "B_θ",
        },
        {
            "name": "v_r",
        },
        {
            "name": "v_φ",
        },
        {
            "name": "v_θ",
        },
        {
            "name": "ρ",
        },
        {
            "name": "B_φ_prime",
        },
    ]

    linestyle = kwargs.pop("line style")
    internal_data = kwargs.pop("internal_data")
    param_angles = np.array(internal_data["angles"])
    params = np.array(internal_data["params"])
    npnot = np.logical_not

    fig, axes = plt.subplots(
        nrows=2, ncols=4, tight_layout=True, sharex=True, **kwargs
    )
    axes.shape = len(param_names)
    for i, settings in enumerate(param_names):
        ax = axes[i]
        pos_params = params[:, i] >= 0
        ax.plot(
            90 - (param_angles[pos_params] * 180 / pi),
            params[pos_params, i], linestyle,
        )
        ax.plot(
            90 - (param_angles[npnot(pos_params)] * 180 / pi),
            - params[npnot(pos_params), i], linestyle,
        )
        ax.set_xlabel("angle from plane (°)")
        ax.set_yscale(settings.get("scale", "log"))
        ax.set_title(settings["name"])
        if settings.get("legend"):
            ax.legend()
        better_sci_format(ax.yaxis)
    return fig


def params_plot_options(parser):
    """
    Add cli arguments for defining derivative plot
    """
    parser.add_argument("--line-style", default=".")


def get_params_plot_args(args):
    """
    Parse plot args
    """
    return {
        "line style": args.get("line_style", "-"),
        "internal_data": args["internal_data"]
    }
