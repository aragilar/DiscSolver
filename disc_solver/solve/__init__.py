# -*- coding: utf-8 -*-
"""
Solver component of Disc Solver
"""
import argparse

import arrow
import logbook
from logbook.compat import redirected_warnings, redirected_logging

from h5py import File

from .config import get_input, step_input
from .solution import create_soln_splitter, solver_generator
from .stepper import (
    binary_searcher, stepper_creator, writer_generator, cleanup_generator
)

from ..file_format import wrap_hdf5_file
from ..logging import logging_options, log_handler
from ..utils import allvars as vars

log = logbook.Logger(__name__)


def solution_main(output_file=None, ismain=True):
    """
    Main function to generate solution
    """
    if ismain:
        parser = argparse.ArgumentParser(description='Solver for DiscSolver')
        parser.add_argument("conffile")
        logging_options(parser)
        args = vars(parser.parse_args())
        conffile = args["conffile"]
    else:
        args = {
            "quiet": True,
        }
        conffile = None
    gen_file_name = True if output_file is None else False
    with log_handler(args), redirected_warnings(), redirected_logging():
        config_inp = get_input(conffile=conffile)
        step_func = step_input(config_inp)
        if gen_file_name:
            output_file = config_inp.label + str(arrow.now()) + ".hdf5"
        with File(output_file) as f:
            soln_file = wrap_hdf5_file(f, version="latest", new=True)
            soln_file.root.config_input = config_inp
            soln_file.root.config_filename = str(conffile)
            writer = writer_generator(soln_file)
            cleanup = cleanup_generator(soln_file, writer)
            binary_searcher(
                solver_generator(), cleanup,
                stepper_creator(
                    writer, step_func,
                    create_soln_splitter("v_Î¸_deriv")
                ),
                config_inp,
            )
