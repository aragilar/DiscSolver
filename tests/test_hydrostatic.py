# -*- coding: utf-8 -*-

from types import SimpleNamespace

import pytest
from pytest import approx

import logbook

import numpy as np
from numpy import sqrt, tan

from disc_solver.float_handling import float_type, FLOAT_TYPE
from disc_solver.utils import ODEIndex
from disc_solver.solve.hydrostatic import ode_system
from disc_solver.analyse.j_e_plot import J_func, E_func

ODE_NUMBER = len(ODEIndex)


@pytest.fixture(scope='module')
def initial_conditions():
    values = SimpleNamespace()
    # These are all arbitrary values, this should not affect the result
    values.params = np.array([7 for i in range(ODE_NUMBER)], dtype=FLOAT_TYPE)
    values.β = float_type(5)/float_type(4)
    values.norm_kepler_sq = float_type(1000)
    values.a_0 = float_type(2)
    values.γ = 0

    values.params[ODEIndex.v_θ] = 0

    # This is slightly off the plane, this should mean we don't get
    # cancellation
    values.angle = float_type(0.01)
    return values


@pytest.fixture(scope='module')
def rhs_func(initial_conditions):
    with logbook.NullHandler().applicationbound():
        rhs, internal_data = ode_system(
            a_0=initial_conditions.a_0,
            norm_kepler_sq=initial_conditions.norm_kepler_sq,
            init_con=initial_conditions.params,
            no_v_deriv=False
        )
    return rhs


@pytest.fixture(scope='module')
def derivs(initial_conditions, rhs_func):
    derivs = np.zeros(ODE_NUMBER)
    with logbook.StderrHandler().applicationbound():
        err = rhs_func(
            initial_conditions.angle,
            initial_conditions.params,
            derivs
        )
    if err != 0:
        raise RuntimeError("rhs failed")
    return derivs


@pytest.fixture(scope='module')
def solution(initial_conditions, derivs):
    values = SimpleNamespace(deriv=SimpleNamespace())

    values.B_r = initial_conditions.params[ODEIndex.B_r]
    values.B_φ = initial_conditions.params[ODEIndex.B_φ]
    values.B_θ = initial_conditions.params[ODEIndex.B_θ]
    values.v_r = initial_conditions.params[ODEIndex.v_r]
    values.v_φ = initial_conditions.params[ODEIndex.v_φ]
    values.v_θ = initial_conditions.params[ODEIndex.v_θ]
    values.ρ = initial_conditions.params[ODEIndex.ρ]
    values.η_O = initial_conditions.params[ODEIndex.η_O]
    values.η_A = initial_conditions.params[ODEIndex.η_A]
    values.η_H = initial_conditions.params[ODEIndex.η_H]
    values.deriv.B_r = derivs[ODEIndex.B_r]
    values.deriv.B_φ = derivs[ODEIndex.B_φ]
    values.deriv.B_θ = derivs[ODEIndex.B_θ]
    values.deriv.v_r = derivs[ODEIndex.v_r]
    values.deriv.v_φ = derivs[ODEIndex.v_φ]
    values.deriv.v_θ = derivs[ODEIndex.v_θ]
    values.deriv.ρ = derivs[ODEIndex.ρ]
    values.deriv.B_φ_prime = derivs[ODEIndex.B_φ_prime]
    values.deriv.η_O = derivs[ODEIndex.η_O]
    values.deriv.η_A = derivs[ODEIndex.η_A]
    values.deriv.η_H = derivs[ODEIndex.η_H]

    B_mag = sqrt(values.B_r**2 + values.B_φ**2 + values.B_θ**2)
    values.norm_B_r, values.norm_B_φ, values.norm_B_θ = (
        values.B_r/B_mag, values.B_φ/B_mag, values.B_θ/B_mag
    )
    return values


def test_continuity(initial_conditions, solution, regtest, test_info, test_id):
    eqn = (
        (
            float_type(5)/float_type(2) - 2 * initial_conditions.β
        ) * solution.v_r + solution.deriv.v_θ + (
            solution.v_θ / solution.ρ
        ) * (
            solution.deriv.ρ - solution.ρ * tan(initial_conditions.angle)
        )
    )
    test_info(eqn)
    regtest.identifier = test_id
    print(eqn, file=regtest)
    assert eqn == approx(0)


def test_solenoid(initial_conditions, solution, regtest, test_info, test_id):
    eqn = solution.deriv.B_θ - (
        - 3/4 * solution.B_r + solution.B_θ * tan(initial_conditions.angle)
    )
    test_info(eqn)
    regtest.identifier = test_id
    print(eqn, file=regtest)
    assert eqn == approx(0)


def test_polar_momentum(
    initial_conditions, solution, regtest, test_info, test_id
):
    eqn = (
        solution.v_r * solution.v_θ / 2 + solution.v_θ * solution.deriv.v_θ +
        tan(initial_conditions.angle) * solution.v_φ ** 2 +
        solution.deriv.ρ / solution.ρ + initial_conditions.a_0 / solution.ρ * (
            (initial_conditions.β - 1) * solution.B_θ * solution.B_r +
            solution.B_r * solution.deriv.B_r +
            solution.B_φ * solution.deriv.B_φ -
            solution.B_φ ** 2 * tan(initial_conditions.angle)
        )
    )
    test_info(eqn)
    regtest.identifier = test_id
    print(eqn, file=regtest)
    assert eqn == approx(0)


def test_polar_induction(
    initial_conditions, solution, regtest, test_info, test_id
):
    eqn = (
        solution.v_θ * solution.B_r -
        solution.v_r * solution.B_θ - (
            solution.deriv.B_r - solution.B_θ * (1 - initial_conditions.β)
        ) * (
            solution.η_O + solution.η_A * (1 - solution.norm_B_φ**2)
        ) - solution.deriv.B_φ * (
            solution.η_H * solution.norm_B_θ +
            solution.η_A * solution.norm_B_r * solution.norm_B_φ
        ) - solution.B_φ * (
            solution.η_H * (
                solution.norm_B_r * (1 - initial_conditions.β) -
                solution.norm_B_θ * tan(initial_conditions.angle)
            ) - solution.η_A * solution.norm_B_φ * (
                solution.norm_B_θ * (1 - initial_conditions.β) +
                solution.norm_B_r * tan(initial_conditions.angle)
            )
        )
    )
    test_info(eqn)
    regtest.identifier = test_id
    print(eqn, file=regtest)
    assert eqn == approx(0)


def test_azimuthal_induction_numeric(
    initial_conditions, derivs, rhs_func, solution, regtest, test_info, test_id
):
    step = float_type(1e-4)
    new_params = initial_conditions.params + derivs * step
    new_angle = initial_conditions.angle + step
    new_derivs = np.zeros(ODE_NUMBER)
    rhs_func(new_angle, new_params, new_derivs)
    dderiv_B_φ_hacked = ((new_derivs - derivs) / step)[ODEIndex.B_φ]
    eqn = dderiv_B_φ_hacked - solution.deriv.B_φ_prime
    test_info(eqn)
    regtest.identifier = test_id
    print(eqn, file=regtest)
    assert eqn == approx(0, abs=5e-12)


def test_E_φ(initial_conditions, solution, regtest, test_info, test_id):
    J_r, J_θ, J_φ = J_func(
        γ=initial_conditions.γ, θ=initial_conditions.angle, B_θ=solution.B_θ,
        B_φ=solution.B_φ, deriv_B_φ=solution.deriv.B_φ,
        deriv_B_r=solution.deriv.B_r,
    )

    E_r, E_θ, E_φ = E_func(
        v_r=solution.v_r, v_θ=solution.v_θ, v_φ=solution.v_φ, B_r=solution.B_r,
        B_θ=solution.B_θ, B_φ=solution.B_φ, J_r=J_r, J_θ=J_θ, J_φ=J_φ,
        η_O=solution.η_O, η_A=solution.η_A, η_H=solution.η_H,
    )

    eqn = E_φ
    test_info(eqn)
    regtest.identifier = test_id
    print(eqn, file=regtest)
    assert eqn == approx(0)
