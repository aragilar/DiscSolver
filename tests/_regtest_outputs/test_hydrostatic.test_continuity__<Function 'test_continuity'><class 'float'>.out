test_hydrostatic.test_continuity__<Function 'test_continuity'><class 'numpy.float64'>.out